### ISSUE TITLE
__new__() cannot be called on a dataclass class

### ISSUE THREAD

**Issue by TeamSpen210** [microsoft/pyright#7567]
**Describe the bug**
Manually calling `__new__()` to bypass `__init__()` works fine on regular classes, but produces an error if the class has a `dataclass_transform` applied. In my case I was using `attrs`, but the same thing happens with `dataclasses`:

**Code or Screenshots**

```python
from dataclasses import dataclass
from typing import Any, Self, reveal_type

class Good:
    def __new__(cls, *args: Any, **kwargs: Any) -> Self:
        return cls()

@dataclass
class Bad:
    pass

reveal_type(Good.__new__)
# Type of "Good.__new__" is "(cls: type[Self@Good], *args: Any, **kwargs: Any) -> Self@Good"
reveal_type(Bad.__new__)
# Type of "Bad.__new__" is "(cls: type[Self@Bad], *args: Any, **kwargs: Any) -> Self@Bad"
Good.__new__(Good)  # Succeeds
Bad.__new__(Bad) 
# error: Argument of type "type[Bad]" cannot be assigned to parameter "cls" of type "type[Self@Bad]" in function "__new__"             
# Type "type[Bad]" cannot be assigned to type "type[Self@Bad]" (reportArgumentType) 
```
[Playground](https://pyright-play.net/?code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAoUSKBATyPwDsBzA408gQTZMANFADK2VMBEhsAN2xxUAfWZFsNGmkx4A4mDAwAXDSinY2YFCVK22AO7WAFGgwiAVHBAcMhqAOFQbm4A1nae3r7%2BAJRQALQAfGISwMZmaVAyCACuIGxQLo5RGgAC8EhatBV4AEJwRiZmRNoaMvKKKizYjnoGAHTWtg5KRa0KyqpdtTD9NvbWRT3TA3NK3fowRVMzg05TMUA)
Replicating the same sort of signature for `__new__()` works fine (as can be seen with `Good` here), as well as omitting it entirely.

**VS Code extension or command-line**
Command line, version 1.1.355, also using the Playground with a version of 1.1.356.

**Comment by Azureblade3808** [comment]
@TeamSpen210 There are still some remaining "Test"s in your code comments which I believe should be "Bad"s.

---

```python
from dataclasses import dataclass
from typing import Any, Self

class C0:
    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...

C1 = dataclass(C0)

reveal_type(C1.__new__)  # Type of "C1.__new__" is "(cls: type[Self@C0], *args: Any, **kwargs: Any) -> Self@C0"
```

Code above should have better explained this issue. Note that `cls` in `C1.__new__` has a type of `type[Self@C0]`.

With some special treatment for dataclasses, `C1` becomes a completely different class from `C0`. Meanwhile, `C0.__new__` is just copied onto `C1` without its `cls` parameter adjusted. Therefore, type mismatch occurs.

**Comment by TeamSpen210** [comment]
Fixed the `Test`, I originally wrote it like that before adding the `Good` example. Indeed a mismatch between the undecorated and decorated type does make sense. But in my example there is no original `Bad.__new__`, that's synthesised by the dataclass implementation. Oddly, if I do `Bad = dataclass(Bad)`, the error vanishes.

**Comment by Azureblade3808** [comment]
This issue appears since v1.335.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
